File: gherkin/php/src/TokenFormatterBuilder.php
<?php

declare(strict_types=1);

namespace Cucumber\Gherkin;

use Cucumber\Gherkin\Parser\Builder;
use Cucumber\Gherkin\Parser\RuleType;

/**
 * @implements Builder<string>
 */
final class TokenFormatterBuilder implements Builder
{
    private TokenFormatter $tokenFormatter;

    /** @var list<string> */
    private $lines = [];

    public function __construct()
    {
        $this->tokenFormatter = new TokenFormatter();
    }

    public function build(Token $token): void
    {
        $this->lines[] = $this->tokenFormatter->formatToken($token);
    }

    public function startRule(RuleType $ruleType): void
    {
    }

    public function endRule(RuleType $ruleType): void
    {
    }

    public function getResult(): string
    {
        // implode at the end is more efficient than repeated concat
        return implode("\n", [...$this->lines, '']);
    }

    public function reset(string $uri): void
    {
    }
}

File: gherkin/dart/lib/src/tokens/TokenFormatterBuilder.dart
import 'package:gherkin/language.dart';
import 'package:gherkin/parser.dart';

import 'TestTokenFormatter.dart';
import 'object.dart';

class TokenFormatterBuilder implements IBuilder<object>
{
  final TestTokenFormatter formatter = TestTokenFormatter();
  final StringBuffer tokensTextBuilder = StringBuffer();

  String getTokensText()
  {
    return tokensTextBuilder.toString();
  }

  @override
  void build(Token token) {
    tokensTextBuilder.writeln(formatter.formatToken(token));
  }

  @override
  void endRule(RuleType ruleType) {
    // nop
  }

  @override
  void reset() {
    // nop
  }

  @override
  object get result => object();

  @override
  void startRule(RuleType ruleType) {
    // nop
  }
}